pydantic/types.py

```diff
@@ -109,116 +109,223 @@
 
     ModelOrDc = Type[Union['BaseModel', 'Dataclass']]
 
+T = TypeVar('T')
 
-class ConstrainedBytes(bytes):
-    strip_whitespace = False
-    min_length: OptionalInt = None
-    max_length: OptionalInt = None
+
+class ConstrainedNumberMeta(type):
```

Comment:
I went a bit further and added a `If TYPE_CHECKING` for almost all custom types.
The diff is big because I added some comments to make the different sections explicit (see the screenshots in description) and moved some class definitions but no extra logic has been changed!

Now almost all custom types work directly with `mypy`.
The generic ones (`conlist`, `conset`) can be handled thanks to the plugin with the `get_dynamic_class_hook` method.I may work on it in a different PR.